packages/vite/src/node/plugins/workerImportMetaUrl.ts

```diff
@@ -117,22 +113,37 @@ export function workerImportMetaUrlPlugin(config: ResolvedConfig): Plugin {
             cleanString,
             index + allExp.length
           )
-          const file = normalizePath(
-            path.resolve(path.dirname(id), rawUrl.slice(1, -1))
-          )
+          const url = rawUrl.slice(1, -1)
+          let file: string | undefined
+          if (url.startsWith('.')) {
+            file = path.resolve(path.dirname(id), url)
+          } else {
+            workerResolver ??= config.createResolver({
+              extensions: [],
+              tryIndex: false,
+              preferRelative: true
+            })
+            file = await workerResolver(url, id)
+            file ??= url.startsWith('/')
+              ? slash(path.join(config.publicDir, url))
+              : slash(path.resolve(path.dirname(id), url))
+          }
 
-          let url: string
+          let builtUrl: string
           if (isBuild) {
             registerWorkersSource(config, id)
-            url = await workerFileToUrl(config, file, query)
+            builtUrl = await workerFileToUrl(config, file, query)
           } else {
-            url = await fileToUrl(cleanUrl(file), config, this)
-            url = injectQuery(url, WORKER_FILE_ID)
-            url = injectQuery(url, `type=${workerType}`)
+            builtUrl = await fileToUrl(file, config, this)
+            builtUrl = injectQuery(builtUrl, WORKER_FILE_ID)
+            builtUrl = injectQuery(builtUrl, `type=${workerType}`)
           }
-          s.overwrite(urlIndex, urlIndex + exp.length, JSON.stringify(url), {
-            contentOnly: true
-          })
+          s.overwrite(
+            urlIndex,
+            urlIndex + exp.length,
+            `new URL(${JSON.stringify(builtUrl)}, self.location)`,
```

Comment:
I don't know why here need to `new URL(${JSON.stringify(builtUrl)}, self.location)`, thank you very much if you could tell me

Reply:
Really I just added this for consistency with the asset loader. The only edge case I can think of where this might have an effect is when the page has a `<base href="https://some-external-origin.com>` tag. Including `self.location` ensures that the asset/worker is resolved relative to the current origin instead of wherever the `<base>` tag is referencing.